src/main/java/com/google/cloud/tools/gradle/appengine/core/DeployExtension.java

```diff
@@ -58,6 +65,65 @@ public DeployExtension(DeployExtension deployExtension) {
     this.stopPreviousVersion = deployExtension.stopPreviousVersion;
     this.version = deployExtension.version;
     this.appEngineDirectory = deployExtension.appEngineDirectory;
+    this.appEngineWebXml = deployExtension.appEngineWebXml;
+  }
+
+  /**
+   * Verifies that project/version properties are pulled correctly from build.gradle and
+   * appengine-web.xml and modifies the properties appropriately.
+   *
+   * @return A copy of the DeployExtension with the corrected properties.
+   */
+  public DeployExtension withPropertiesFromAppEngineWebXml()
+      throws AppEngineException, IOException, SAXException {
+    DeployExtension deployExtensionCopy = new DeployExtension(this);
+
+    if (appEngineWebXml == null) {
+      return deployExtensionCopy;
+    }
+
+    AppEngineDescriptor appengineWebXmlDoc =
+        AppEngineDescriptor.parse(new FileInputStream(appEngineWebXml));
+    String xmlProject = appengineWebXmlDoc.getProjectId();
+    String xmlVersion = appengineWebXmlDoc.getProjectVersion();
+
+    // Verify that project and version are set somewhere
+    if (project == null && xmlProject == null || version == null && xmlVersion == null) {
+      throw new GradleException(
+          "appengine-plugin does not use gcloud global project state. Please configure the "
+              + "application ID and version in your build.gradle or appengine-web.xml.");
+    }
+
+    // Check system property
+    boolean readAppEngineWebXml =
+        System.getProperty("deploy.read.appengine.web.xml") != null
+            && System.getProperty("deploy.read.appengine.web.xml").equals("true");
+    if (readAppEngineWebXml) {
+      // Use properties from appengine-web.xml if not also set in build.gradle
+      if (project != null && xmlProject != null || version != null && xmlVersion != null) {
+        throw new GradleException(
```

Comment:
I wonder if this should happen even if the xml properties are empty, i.e. if the system property is set and there's something in the build.gradle, error out, regardless of whether or not there are values to overwrite it with from the xml.